### PR TITLE
Fix SIGSEGV when running `newsboat -r`

### PR DESCRIPTION
**Pull request by Minoru**
While investigating #1163, I found that `newsboat -r` segfaults for me with the following backtrace:

```
#0  0x000056442a4a8976 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::size (this=0x98) at /usr/include/c++/10/bits/basic_string.h:2833
#1  std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_check (__s=0x56442a705270 "basic_string::substr", __pos=0, this=0x98) at /usr/include/c++/10/bits/basic_string.h:312
#2  std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::substr (__n=6, __pos=0, this=0x98) at /usr/include/c++/10/bits/basic_string.h:2835
#3  newsboat::RssFeed::is_query_feed (this=0x0) at include/rssfeed.h:142
#4  newsboat::Reloader::reload (this=0x56442bd73690, pos=1, max=567, unattended=<optimized out>, easyhandle=0x7f031bffe978) at src/reloader.cpp:67
#5  0x000056442a4a994b in newsboat::Reloader::reload_range (this=0x56442bd73690, start=<optimized out>, end=<optimized out>, size=567, unattended=false) at src/reloader.cpp:253
#6  0x00007f03249a7c60 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#7  0x00007f0324ecfea7 in start_thread (arg=<optimized out>) at pthread_create.c:477
#8  0x00007f03246b1eaf in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

Note that `this` pointer is zero in frame 3.

I failed to create a sufficiently small reproducible example; looks like one needs my cache file to trigger this, and also use more than one reload thread.

I also failed to pinpoint the exact reason for the segfault. However, while reading the code I noticed that:

1. `FeedContainer` stores a vector of feeds, and that vector *is public*;
2. the vector is protected by a private `FeedContainer::feeds_mutex`, which is used in all relevant methods of `FeedContainer`;
3. `Controller` has its own `feeds_mutex` and uses it every time it accesses `FeedContainer::feeds`.

I guess this is a leftover from a refactoring we made a few years back, when we split `Controller` into smaller classes — that's when we introduced `FeedContainer` (https://github.com/newsboat/newsboat/pull/216).

This fixes the segfault for me. If `newsboat -r` segfaults for someone else, please check out these changes and see if they fix it for you, too.

Reviews welcome! I'll merge this in three days, provided the CI is green and no issues are raised.

### CODE REVIEW

**Comment by Minoru** [comment]
Oh, turns out we already had an issue about making `FeedContainer::feeds` private: https://github.com/newsboat/newsboat/issues/1027. Merging this PR will close that issue.

File: src/feedcontainer.cpp [hunk 1]
```diff
@@ -289,4 +289,13 @@ unsigned int FeedContainer::unread_item_count() const
 	return unread_guids.size();
 }
 
+void FeedContainer::replace_feed(unsigned int pos,
+	std::shared_ptr<RssFeed> feed)
+{
+	assert(pos < feeds.size());
```
Comment (dennisschagt) [inline src/feedcontainer.cpp:291]: It looks somewhat strange to have the assert before locking `feeds_mutex`.
I expect it is possible for the controller to call `FeedContainer::set_feeds(...)` in a way which lowers the amount of feeds (in-between the assert and the locking of the `feeds_mutex`).

File: src/controller.cpp [hunk 2]
```diff
@@ -594,7 +594,7 @@ void Controller::mark_article_read(const std::string& guid, bool read)
 
 void Controller::mark_all_read(unsigned int pos)
 {
-	if (pos < feedcontainer.feeds.size()) {
+	if (pos < feedcontainer.feeds_size()) {
```
Comment (dennisschagt) [inline src/controller.cpp:592]: This function might contain a race condition.

The validity of the `pos` value is checked here but the number of feeds might be changed before the two uses of `pos`:
- `const auto feed = feedcontainer.get_feed(pos);`
- `feedcontainer.mark_all_feed_items_read(pos);`

**Review by dennisschagt — COMMENTED** [review]
Looks good.

I found two possible race conditions, not completely sure if those are actual issues.

**Comment by Minoru** [comment]
Both are definitely issues, thanks for spotting them! I'm working on the fixes now.

File: src/feedcontainer.cpp (hunk 1 above)
Comment (Minoru) [reply src/feedcontainer.cpp]: Fixed by https://github.com/newsboat/newsboat/pull/1166/commits/d570f32610780e975d61bccbe0dc60bd5d8f5341.

File: src/controller.cpp (hunk 2 above)
Comment (Minoru) [reply src/controller.cpp]: Fixed by the last three commits I just pushed.

Comment (dennisschagt) [reply src/controller.cpp]: Looks good, thanks for fixing!

**Review by dennisschagt — APPROVED** [review]
Looks good!

**Comment by Minoru** [comment]
Thanks! I'm squashing the fixup and force-pushing; will merge once all the CI checks pass.

**Pull request merged**